crates/sui-core/src/transaction_input_checker.rs

```diff
@@ -74,7 +74,28 @@ pub async fn check_transaction_input(
     transaction.validity_check(epoch_store.protocol_config())?;
     let input_objects = transaction.input_objects()?;
     let objects = store.check_input_objects(&input_objects)?;
-    let gas_status = get_gas_status(&objects, epoch_store, transaction).await?;
+    let gas_status = get_gas_status(&objects, transaction.gas(), epoch_store, transaction).await?;
+    let input_objects = check_objects(transaction, input_objects, objects)?;
+    Ok((gas_status, input_objects))
+}
+
+pub async fn check_transaction_input_no_gas(
+    store: &AuthorityStore,
+    epoch_store: &AuthorityPerEpochStore,
+    transaction: &TransactionData,
+    gas_object: Object, // fake gas object
```

Comment:
The fact that a function that has `no_gas` in the name accepts a `gas_object` + the fact that it presumably has some very specific expectations around that gas object (that it's fake, it has the max possible balance, etc), makes me think that this function should create the `gas_object` internally to itself, thoughts?

Reply:
I actually like it and moving it there